File: solutions/0053_Maximum_Subarray/dp.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        dp = [0] * len(nums)
        dp[0], best = nums[0], nums[0]
        for i in range(1, len(nums)):
            dp[i] = max(dp[i-1] + nums[i], nums[i])
            best = max(dp[i], best)
        return best

File: solutions/0343_Integer_Break/dp.java
class Solution {
    public int integerBreak(int n) {
        if (n == 1) return 1;
        int[] dp = new int[n + 1];
        dp[1] = 1;
        for (int i = 1; i <= n; ++i) {
            int best = 0;
            for (int j = 1; j <= i / 2; ++j) {
                best = Math.max(best,
                                Math.max(dp[j], j) * Math.max(i - j, dp[i - j]));
            }
            dp[i] = best;
        }
        return dp[n];
    }
}
